ragna/deploy/_ui/components/metadata_filters_builder.py

```diff
@@ -0,0 +1,307 @@
+            key=self.key, value=self.convert_value(self.value)
+        )
+
+
+class MetadataFiltersBuilder(pn.viewable.Viewer):
+    metadata_filters = param.List([])
+
+    def __init__(self, **params):
+        super().__init__(**params)
+
+        self.add_filter_button = pn.widgets.ButtonIcon(
+            icon="circle-plus", width=25, height=25
+        )
+        self.add_filter_button.on_click(self.did_click_add_filter_button)
+
+        self.delete_buttons = []
+
+        # dummy_row = FilterRow(key="document_name", operator="==", value="applications.md")
+        self.metadata_filters.append(self.empty_filter())
+
+    def empty_filter(self):
+        return FilterRow(
+            key="",
+            operator="",
+            value="",
+            on_delete_callback=self.delete_metadata_filter,
+        )
+
+    def did_click_add_filter_button(self, event):
+        if len(self.metadata_filters) > 0 and not self.metadata_filters[-1].is_empty():
+            new_metadata_filter_row = self.empty_filter()
+            self.metadata_filters = self.metadata_filters + [new_metadata_filter_row]
+
+    def delete_metadata_filter(self, event):
+        print("should delete filter : ", event)
+
+        filter_to_remove = None
+        for filter in self.metadata_filters:
+            if event.obj == filter.delete_button:
+                filter_to_remove = filter
+                break
+
+        if filter_to_remove is None:
+            return
+
+        new_metadata_filters = [f for f in self.metadata_filters if f != filter]
+        if len(new_metadata_filters) == 0:
+            new_metadata_filters.append(self.empty_filter())
+
+        self.metadata_filters = new_metadata_filters
+
+    @pn.depends("metadata_filters")
+    def render_metadata_filters_rows(self):
+        return pn.Column(*self.metadata_filters)
+
+    def get_metadata_filters(self):
+        if len(self.metadata_filters) == 0:
+            return None
+        elif len(self.metadata_filters) == 1:
+            if self.metadata_filters[0].is_empty():
+                return None
+            return self.metadata_filters[0].get_metadata_filter().to_primitive()
+        else:
+            return MetadataFilter.and_(
+                [
+                    f.get_metadata_filter()
+                    for f in self.metadata_filters
+                    if f.get_metadata_filter() is not None
+                ]
+            ).to_primitive()
+
+    def validate(self):
+        result = True
+        for filter in self.metadata_filters:
+            # If the last filter is empty, we do not want to validate it.
+            # This allows to have only one empty filter, to question on the whole corpus.
+            if filter == self.metadata_filters[-1] and filter.is_empty():
+                continue
+
+            if not filter.validate():
+                result = False
+                # Do not break, we want to call validate() on every filters
```

Comment:
Why?

Reply:
Let's take this example : 
![Capture d’écran 2024-08-14 à 09 44 14](https://github.com/user-attachments/assets/423e673c-f9ba-4b6e-9cc5-a76e5c3671d9)

We have 2 empty fields. We want to validate all of them, and not stop after the first, to highlight in red all the invalid ones : 
![Capture d’écran 2024-08-14 à 09 44 41](https://github.com/user-attachments/assets/9afd3b64-beb9-469f-bb22-328c8389367c)

Reply:
In that case the variable naming is confusing. So `filter` is not a row, but rather one element inside the row? Maybe name it `filter_component` than?